### PR TITLE
Sorr task538 centralize input validation when running etl scripts

### PR DESCRIPTION
**Pull request by Sameep2808**
Task #538 @DanilYachmenev I have created the new file and transferred the functions as per the issue

### CODE REVIEW

File: im_v2/common/data/qa/validate_input_args.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env python
```
Comment (DanilYachmenev) [inline im_v2/common/data/qa/validate_input_args.py:1]: remove
this one is needed only for scripts, and this is a lib file

File: im_v2/common/data/qa/validate_input_args.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+"""
+Centralizing checking logical dependencies between values of various arguments
```
Comment (DanilYachmenev) [inline im_v2/common/data/qa/validate_input_args.py:3]: -> `Centralize ...`

File: im_v2/common/data/qa/validate_input_args.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import im_v2.crypto_chassis.data.extract.extractor as imvccdexex
+
+
+def validate_vendor_arg(
+    vendor: str, args: argparse.Namespace
+) -> ivcdexex.Extractor:
+    """
+    Function to validate vendors arguments for contract type.
+
+    :param vendor: string to get the vendor
```
Comment (DanilYachmenev) [inline im_v2/common/data/qa/validate_input_args.py:25]: -> `:param vendor: vendor name`

File: im_v2/common/data/qa/validate_input_args.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    :param args: args to get the contract type
+    """
```
Comment (DanilYachmenev) [inline im_v2/common/data/qa/validate_input_args.py:27]: add also `:return: vendor data extractor`

File: im_v2/common/data/qa/validate_input_args.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import argparse
+
+import helpers.hdbg as hdbg
+import im_v2.binance.data.extract.extractor as imvbdexex
+import im_v2.ccxt.data.extract.extractor as imvcdexex
+import im_v2.common.data.extract.extractor as ivcdexex
+import im_v2.crypto_chassis.data.extract.extractor as imvccdexex
+
+
+def validate_vendor_arg(
```
Comment (DanilYachmenev) [inline im_v2/common/data/qa/validate_input_args.py:19]: @jsmerix should we maybe give it a different name?
it seems that this function now is rather about getting extractor than just validation

**Review by DanilYachmenev — COMMENTED** [review]
sg, couple moments

File: im_v2/common/data/qa/validate_input_args.py (hunk 5 above)
Comment (jsmerix) [reply im_v2/common/data/qa/validate_input_args.py:18]: If this module is for validation then only validations (assertions) belong here. 
This function should only check:
- if vendor belongs to one of the three supported
- If `universe_part` is included when  the vendor arg is `crypto_chassis`

Having a function which returns the correct instance is also useful but should belong to a different module, okay for me to file a follow-up.

**Review by jsmerix — CHANGES_REQUESTED** [review]

File: im_v2/common/data/qa/validate_input_args.py (hunk 5 above)
Comment (DanilYachmenev) [reply im_v2/common/data/qa/validate_input_args.py:18]: git it, thx

```
def validate_vendor_arg(vendor: str, args: argparse.Namespace) -> None:
    """
    Validate vendor args.
    
    :param vendor: vendor name
    :param args: ETL script args
    """
    hdbg.dassert_in(vendor, ["binance", "ccxt", "crypto_chassis"], msg=f"Vendor {vendor} is not supported.")
    if vendor == "crypto_chassis":
        if not args.get("universe_part"):
            raise RuntimeError(
                f"--universe_part argument is mandatory for {vendor}"
            )
```

@Sameep2808 lets implement the changes here
update the script accordingly (run this function first and then init extractor)

**Comment by Sameep2808** [comment]
@DanilYachmenev did the required changes you can have a look now

**Review by jsmerix — APPROVED** [review]

File: im_v2/common/data/extract/download_bulk.py [hunk 6]
```diff
@@ -92,8 +90,6 @@ def _run(args: argparse.Namespace) -> None:
             # TODO(Vlad): Temporary stick to daily data for Binance.
             time_period=imvbdexex.BinanceNativeTimePeriod.DAILY,
         )
-    else:
-        hdbg.dfatal(f"Vendor {vendor} is not supported.")
```
Comment (DanilYachmenev) [inline im_v2/common/data/extract/download_bulk.py:96]: lets keep these lines
`if` clauses should always be completed with `else`
https://github.com/sorrentum/sorrentum/blob/master/docs/Coding_Style_Guide.md#make-if-elif-else-complete

**Review by DanilYachmenev — COMMENTED** [review]
1 nit and lgtm

**Comment by Sameep2808** [comment]
> lets keep these lines `if` clauses should always be completed with `else` https://github.com/sorrentum/sorrentum/blob/master/docs/Coding_Style_Guide.md#make-if-elif-else-complete

@DanilYachmenev I just thought that would have been a DRY with [this](https://github.com/sorrentum/sorrentum/pull/552/files#:~:text=else%3A,not%20supported.%22)

**Comment by Sameep2808** [comment]
> lets keep these lines `if` clauses should always be completed with `else` https://github.com/sorrentum/sorrentum/blob/master/docs/Coding_Style_Guide.md#make-if-elif-else-complete

I thought that `else` would be DRY with [this](https://github.com/sorrentum/sorrentum/pull/552/files#:~:text=hdbg.dassert_in,not%20supported.%22%2C)

**Comment by gpsaggese** [comment]
> > lets keep these lines `if` clauses should always be completed with `else` https://github.com/sorrentum/sorrentum/blob/master/docs/Coding_Style_Guide.md#make-if-elif-else-complete
> 
> I thought that `else` would be DRY with [this](https://github.com/sorrentum/sorrentum/pull/552/files#:~:text=hdbg.dassert_in,not%20supported.%22%2C)

This is a valid point, but in this case the DRY principle succumbs to keeping the code maintainable. 

I would write the code like:

```python
    hdbg.dassert_in(
        vendor,
        ["binance", "ccxt", "crypto_chassis"],
        msg=f"Vendor {vendor} is not supported.",
    )
    if vendor == "crypto_chassis":
        if not args.get("universe_part"):
            raise RuntimeError(
                f"--universe_part argument is mandatory for {vendor}"
            )
    elif vendor == "ccxt":
        # Do something or raise NotImplementedError
    elif vendor == "crypto_chassis":
        # Do something or raise NotImplementedError
    else:
        raise ValueError("Invalid error vendor='${vendor}'")
```

Assume that you add another vendor, then we centralize the assertion and add a new vendor to assertion.
Then we need to go and update all the if-then-else everywhere.
Surely we will miss one. With your solution you will get a silent error (return a `None`), the solution above will at least assert and you can find that there is a missing branch.

If the code in each branch I would actually remove the `dassert`.
The `dassert` is mainly to fail quickly, but it is redundant. So you need a way to keep the redundancy in sync, hence the `else`.

From docs/Reading_List.md
```
# Coding

- [The Pragmatic Programmer](https://www.amazon.com/Pragmatic-Programmer-Journeyman-Master/dp/020161622X)
  - Aka the Black Book
  - Reading (and really internalizing) this book is equivalent to the experience
    you can gain by coding for 20 years
  - It will change your life

- [Code Complete: A Practical Handbook of Software Construction, Second
  Edition](https://www.amazon.com/Code-Complete-Practical-Handbook-Construction/dp/0735619670)
  - Likely the only decent product ever made by Micro$oft
```

**Comment by Sameep2808** [comment]
> Assume that you add another vendor, then we centralize the assertion and add a new vendor to assertion.
> Then we need to go and update all the if-then-else everywhere.
> Surely we will miss one. With your solution you will get a silent error (return a `None`), the solution above will at least assert and you can find that there is a missing branch.

> If the code in each branch I would actually remove the `dassert`.
> The `dassert` is mainly to fail quickly, but it is redundant. So you need a way to keep the redundancy in sync, hence the `else`.

@gpsaggese Thanks for clarifying I understood why the `else` should be implemented. Will keep this in mind for the future implementation.

Just one clarification. 

> ```python
>     elif vendor == "ccxt":
>         # Do something or raise NotImplementedError
>     elif vendor == "crypto_chassis":
>         # Do something or raise NotImplementedError
>     else:
>         raise ValueError("Invalid error vendor='${vendor}'")
> ```

The above part is already implemented [here](https://github.com/sorrentum/sorrentum/blob/edbd2a2133ace3e5e250a93b6d3f06826f80b64a/im_v2/common/data/extract/download_bulk.py#L82C9-L96C58) just after this function is called. As discussed earlier in this issue this function that we are implementing is just for validation purpose. And hence the remaining else if were transferred back to `im_v2/common/data/extract/download_bulk.p`. But do let me know if you want me to transfer it back and I'll make the needed changes.

**Review by DanilYachmenev — COMMENTED** [review]
lgtm

**Comment by gpsaggese** [comment]
Let's make the changes for this bug and I've filed a follow up bug #557

**Review by jsmerix — APPROVED** [review]

**Pull request merged**